source/simulator/core.cc

```diff
@@ -870,125 +870,157 @@ namespace aspect
+            coupling[x.velocities[d]][
+              introspection.variable("compaction pressure").first_component_index] = DoFTools::always;
+            coupling[introspection.variable("compaction pressure").first_component_index]
+            [x.velocities[d]]
+              = DoFTools::always;
+            coupling[x.velocities[d]]
+            [introspection.variable("fluid pressure").first_component_index]
+              = DoFTools::always;
+            coupling[introspection.variable("fluid pressure").first_component_index]
+            [x.velocities[d]]
+              = DoFTools::always;
           }
+
+        coupling[introspection.variable("fluid pressure").first_component_index]
+        [introspection.variable("fluid pressure").first_component_index]
+          = DoFTools::always;
+        coupling[introspection.variable("compaction pressure").first_component_index]
+        [introspection.variable("compaction pressure").first_component_index]
+          = DoFTools::always;
       }
+    else
+      {
+        // The AMG matrix based solver: all velocities couple with all velocities,
+        // pressure couples with all velocities and the other way around,
+        // and pressures only couple with themselves for equal order elements
+        for (unsigned int c=0; c<dim; ++c)
+          for (unsigned int d=0; d<dim; ++d)
+            coupling[x.velocities[c]][x.velocities[d]] = DoFTools::always;
 
-    Table<2,DoFTools::Coupling> coupling (introspection.n_components,
-                                          introspection.n_components);
-    coupling.fill (DoFTools::none);
+        for (unsigned int d=0; d<dim; ++d)
+          {
+            coupling[x.velocities[d]][x.pressure] = DoFTools::always;
+            coupling[x.pressure][x.velocities[d]] = DoFTools::always;
+          }
 
-    // determine which blocks should be fillable in the matrix.
-    // note:
-    // - all velocities couple with all velocities
-    // - pressure couples with all velocities and the other way
-    //   around
-    // - temperature only couples with itself
-    // - compositional fields only couple with themselves
-    // - additionally, in models with melt transport fluid pressure
-    //   and compaction pressures couple with themselves
-    {
-      const typename Introspection<dim>::ComponentIndices &x
-        = introspection.component_indices;
+        // For equal-order interpolation, we need a stabilization term
+        // in the bottom right of Stokes matrix. Make sure we have the
+        // necessary entries.
+        if (parameters.use_equal_order_interpolation_for_stokes == true)
+          coupling[x.pressure][x.pressure] = DoFTools::always;
+      }
 
-      for (unsigned int c=0; c<dim; ++c)
-        for (unsigned int d=0; d<dim; ++d)
-          coupling[x.velocities[c]][x.velocities[d]] = DoFTools::always;
+    // Now determine the coupling for the temperature block
+    coupling[x.temperature][x.temperature] = DoFTools::always;
```

Comment:
can you move remove this line and put it inside the if below by inverting the logic for the if?